taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
+        in 2018 to $500,000 and to continue indexing it in subsequent
+        years as in current-law policy, the YEAR:MODS dictionary would
+        be as follows::
+
+            {2018: {"_SS_Earnings_c":[500000]}}
+
+        But to raise the maximum taxable earnings in 2018 to $500,000
+        without any indexing in subsequent years, the YEAR:MODS
+        dictionary would be as follows::
+
+            {2018: {"_SS_Earnings_c":[500000], "_SS_Earnings_c_cpi":False}}
+
+        And to raise in 2019 the starting AGI for EITC phaseout for
+        married filing jointly filing status (which is a two-dimensional
+        policy parameter that varies by the number of children from zero
+        to three or more and is inflation indexed), the YEAR:MODS dictionary
+        would be as follows::
+
+            {2019: {"_EITC_ps_MarriedJ":[[8000, 8500, 9000, 9500]]}}
+
+        Notice the pair of double square brackets around the four values
+        for 2019.  The one-dimensional parameters above require only a pair
+        of single square brackets.
+        """
+        # check YEAR value in the single YEAR:MODS dictionary parameter
+        if not isinstance(year_mods, dict):
+            msg = 'year_mods is not a dictionary'
+            raise ValueError(msg)
+        if len(year_mods.keys()) != 1:
+            msg = 'year_mods dictionary must contain a single YEAR:MODS pair'
+            raise ValueError(msg)
+        year = list(year_mods.keys())[0]
+        if not isinstance(year, int):
+            msg = 'YEAR in the YEAR:MODS pair in year_mods is not an integer'
+            raise ValueError(msg)
+        if year != self.current_year:
+            msg = 'YEAR={} in year_mods is not equal to current_year={}'
+            raise ValueError(msg.format(year, self.current_year))
+
+        # implement reform provisions included in the single YEAR:MODS pair
+        num_years_to_expand = (self.start_year + self.num_years) - year
+        inf_rates = [self._inflation_rates[(year - self.start_year) + i]
+                     for i in range(0, num_years_to_expand)]
+        paramvals = self._vals
+        for name, values in year_mods[year].items():
+            # determine inflation indexing status of parameter with name
+            if name.endswith('_cpi'):
+                continue
+            if name in paramvals:
+                default_cpi = paramvals[name].get('cpi_inflated', False)
+            else:
+                default_cpi = False
+            cpi_inflated = year_mods[year].get(name + '_cpi', default_cpi)
+            # set post-reform values of parameter with name
+            nval = expand_array(values,
+                                inflate=cpi_inflated,
+                                inflation_rates=inf_rates,
+                                num_years=num_years_to_expand)
+            cval = getattr(self, name)
+            cval[(self.current_year - self.start_year):] = nval
+            setattr(self, name, cval)
```

Comment:
It's correct to note that `cval` is not a copy (deep or shallow) of the attribute (whose name is the contents of `name`). Sometimes it's helpful to compare variables in Python to "references" in other languages, say C++, although that analogy falls apart when you start to push it too hard. I've found that the best way to think about variables in Python is just to think of them as names, or even just "name tags" that we stick on things, the "things" being values.